### PR TITLE
Jama matrix module with unit tests

### PR DESCRIPTION
**Pull request by ct-clmsn**
Included is a direct port of the JAMA java matrix library developed by NIST and Mathworks to Chapel. The source can be found here: http://math.nist.gov/javanumerics/jama/ (JAMA is released as public domain software)

The module currently supports single node computation. Versions of these algorithms and operations targeting distributed arrays (where applicable) would be ideal (place in the TODO column). The module does not currently leverage the LAPACK module; replacing parts of this code with appropriate calls to LAPACK would be a great next step.

Unit tests are included with this PR. The unit tests were derived from the existing JAMA java unit tests.

The module supports the following matrix operations:

Object Manipulation 	
constructors
set elements
get elements
copy
clone

Elementary Operations 	
addition
subtraction
multiplication
scalar multiplication
element-wise multiplication
element-wise division
unary minus
transpose
norm

Decompositions 	
Cholesky
LU
QR
SVD
symmetric eigenvalue
nonsymmetric eigenvalue

Equation Solution 	
nonsingular systems
least squares

Derived Quantities 	
condition number
determinant
rank
inverse
pseudoinverse

### CODE REVIEW

**Comment by ct-clmsn** [comment]
What advice can ya'll offer to get this PR passing the travis checks?

**Comment by ronawho** [comment]
@ct-clmsn It looks like travis is currently failing because of a missing copyright the the jama.chpl module.

Should be pretty easy to fix that issue by running:
```
$CHPL_HOME/util/buildRelease/add_license_to_sources.py $CHPL_HOME/modules/standard/Jama.chpl
```

**Comment by ct-clmsn** [comment]
@ronawho thanks for the help! modifications made and commited!

**Comment by Kyle-B** [comment]
Hi @ct-clmsn, sorry it took us so long to get back to you on this.

Over all, things look pretty good. Github wont let me comment inline on `Jama.chpl`, so I'll make some notes here. These are not major issues, but I think they'd be nice to fix before merging:


* [Jama.chpl:713](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L713): Anywhere there is a `for (i,j) in {...}` see if you can reuse an existing domain. In this case, `{1..n, 1..n} is going to create a new domain, this overhead could be avoided by directly iterating over V's domain like so:
    ```chpl
    for (i, j) in V.domain { ... }
    ```

* [Jama.chpl:1611](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L1611): I think this could just be a whole array assign? (See also: [1625](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L1625), [1900](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L1900), [1919](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L1919), [1287](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L1287), maybe elsewhere?) Doing whole array assigns will be faster in general, mostly by avoiding creating a new domain as mentioned above.

* [Jama.chpl:26](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L26): Could long lines like this please be wrapped? We try to keep most things around 80 columns.

* [Jama.chpl:202](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L202): There is a decent amount of commented out Java in this file, I think it'd be good to remove all of it. (see also: [1500](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L1500), [2674](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L2674))

* Jama.chpl: It would be nice to convert the Javadoc style comments to work with [chpldoc](http://chapel.cray.com/docs/1.12/technotes/chpldoc.html). If you add `Jama.chpl` to the `MODULES_TO_DOCUMENT` list in `CHPL_HOME/modules/Makefile`, the documentation will be automatically generated and placed online, alongside the rest of our standard modules.

Minor things, mostly stylistic:

* [Jama.chpl:2012 & 2117](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L2012): This seems like it would be a good place to use the `this` function (`proc this(i:int, j: int) ref` which allows using the `M[i,j] = x` syntax). But I'd understand if you want to make this library match the Java version.

* [Jama.chpl:2304](https://github.com/ct-clmsn/chapel/blob/jama/modules/standard/Jama.chpl#L2304): I this and the other similar routines could be implemented using promoted array operations:
    ```chpl
    C = A + B.A
    ```

I've not reviewed the details of most of the solvers and related routines, but I'm assuming they are fine.

File: test/modules/standard/jama/TestMatrix.chpl [hunk 1]
```diff
@@ -0,0 +1,403 @@
+use Jama;
+use Math;
+
+/** TestMatrix tests the functionality of the Jama Matrix class and associated decompositions.
+<P>
+Run the test from the command line using
+<BLOCKQUOTE><PRE><CODE>
```
Comment (Kyle-B) [inline test/modules/standard/jama/TestMatrix.chpl]: These comments don't need to work with chpldoc, but it would be nice to strip out the <FOO> tags.

Comment (ct-clmsn) [inline test/modules/standard/jama/TestMatrix.chpl]: not a problem - will go back and make the modifications you've outlined!

Comment (ct-clmsn) [inline test/modules/standard/jama/TestMatrix.chpl]: @Kyle-B for whole array assignment is the preference to do 'p(..) = piv(..)' or just a simple 'p = piv'?(Jama.chpl:1612)

Comment (Kyle-B) [inline test/modules/standard/jama/TestMatrix.chpl]: We usually just do `p = piv` for whole array assign.

Comment (ct-clmsn) [inline test/modules/standard/jama/TestMatrix.chpl]: Would ya'll mind if this code module is renamed to something more recognizable? Maybe DenseLinearAlgebra or LinearAlgebra?

Comment (bradcray) [inline test/modules/standard/jama/TestMatrix.chpl]: What do you think about a blended name like LinearAlgebra-Jama.chpl?  My slight hesitancy in calling it something completely generic is that if we change the "default" linear algebra interface down the road, we break code.  Whereas if, over time, we say "Yeah, this is *the* blessed linear algebra package for Chapel!" then we can rename this and just have people update a 'use' clause (or continue to support the alias for backwards compatibility).

**Comment by bradcray** [comment]
@ct-clmsn , @Kyle-B -- just checking in to see where this PR stands (e.g., who owns it currently?), as I've lost track.  Thanks.

**Comment by bradcray** [comment]
Following up on my previous comment, @Kyle-B asserts that it's in Chris's court.  Apparently GitHub doesn't allow us to assign a PR to an external contributor (awwww... :(  ), so I'm assigning it to Kyle as a placeholder just so that nobody on the team worries that it's in limbo (as I regularly do).

**Comment by ct-clmsn** [comment]
@Kyle-B & @bradcray Apologies for letting this drop off my radar. Went through and got several code suggested edits completed, hopefully in time for the upcoming release. Do ya'll want more polish applied? Thanks in advance!

**Comment by lydia-duncan** [comment]
Hi @ct-clmsn - the failure seems to be related to the copyright - I'm guessing it still says 2015 instead of 2016?

File: test/modules/standard/jama/TestMatrix.chpl (hunk 1 above)
Comment (lydia-duncan) [inline test/modules/standard/jama/TestMatrix.chpl]: It looks like the tags didn't get removed yet.

File: test/modules/standard/jama/TestMatrix.chpl [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+ java Jama.test.TestMatrix
```
Comment (lydia-duncan) [inline test/modules/standard/jama/TestMatrix.chpl]: This command also seems to refer to the java version.  I'm guessing that doesn't apply to the Chapel file?

File: test/modules/standard/jama/TestMatrix.chpl [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+   proc main () {
+
+      var A,B,C,Z,O,I,R,S,X,SUB,M,T,SQ,DEF,SOL : Matrix;
+      // Uncomment this to test IO in a different locale.
+      // Locale.setDefault(Locale.GERMAN);
+      var errorCount=0;
+      var warningCount=0;
+      var tmp, s : real;
+
+      //double[] columnwise = {1.,2.,3.,4.,5.,6.,7.,8.,9.,10.,11.,12.};
```
Comment (lydia-duncan) [inline test/modules/standard/jama/TestMatrix.chpl]: This is commented out Java, correct?  I think it'd be nice to clean this up and leave a reference to the original java file, to give credit where credit is due.

File: test/modules/standard/jama/TestMatrix.chpl [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+      T = new Matrix(tvals);
+      T = A.transpose();
+         writeln(A.transpose(),T);
+check(A.transpose(),T);
+         writeln("transpose...","");
+
+      A.transpose();
+check(A.norm1(),columnsummax);
+         writeln("norm1...","");
+check(A.normInf(),rowsummax);
```
Comment (lydia-duncan) [inline test/modules/standard/jama/TestMatrix.chpl]: The indentation is a little weird here

File: test/modules/standard/jama/TestMatrix.chpl [hunk 5]
```diff
@@ -0,0 +1,403 @@
+use Jama;
```
Comment (lydia-duncan) [inline test/modules/standard/jama/TestMatrix.chpl]: While it is nice to have the test code near what it tests, that differs pretty significantly from our usual style.  Maybe consider moving this and the .good file under a new folder test/modules/standard/jama?  That will allow us to run it nightly

**Comment by lydia-duncan** [comment]
The large comment at the top of the Jama.chpl file seems Java specific (and as part of that claims a different copyright than what is present in the first line comment of the file).  I'd be tempted to tweak it, so that it is more Chapel specific, while still giving credit to the original source

**Comment by lydia-duncan** [comment]
It'd be nice to modify the documentation comments so that they match our chpldoc style (see modules/standard/IO.chpl and FileSystem.chpl for examples of this).  I wouldn't gate this PR on it, since the code should get in before feature freeze next Monday, but it'd be a shame for the documentation updates to not make it before doc freeze the following Monday.

**Comment by lydia-duncan** [comment]
It looks like some of Kyle's original comments are still unresolved, what's the status on those?

File: test/modules/standard/jama/TestMatrix.chpl (hunk 2 above)
Comment (ct-clmsn) [inline test/modules/standard/jama/TestMatrix.chpl]: @lydia-duncan yep. just went through trying to remove some of those artifacts - thanks for double checking the testing code that was ported over!

File: test/modules/standard/jama/TestMatrix.chpl (hunk 3 above)
Comment (ct-clmsn) [inline test/modules/standard/jama/TestMatrix.chpl]: Definitely - i've added a reference to the original jama page and explained the code is a port from the original jama package.

File: test/modules/standard/jama/TestMatrix.chpl (hunk 4 above)
Comment (ct-clmsn) [inline test/modules/standard/jama/TestMatrix.chpl]: @lydia-duncan fixed!

File: test/modules/standard/jama/TestMatrix.chpl (hunk 5 above)
Comment (ct-clmsn) [inline test/modules/standard/jama/TestMatrix.chpl]: gotcha - where'd you want me to move the source to?

**Comment by ct-clmsn** [comment]
Jama.chpl:713, Jama.chpl:1611 - domain-reuse - I modified most of the nested loops that didn't have a range dependencies based on an outer loop. I'll double check, if I missed a couple or if some immediately jump out could you post them here (line num)? :+1: 
Jama.chpl:1611 - assignments - tried to clean those out. in some cases, i need to do it b/c it was a row-copy or column-copy from a matrix into a vector. I'll double check and if you see some of these please post! :+1: 
Jama.chpl:26 - long lines - tried to keep the lines to ~80 chars. probably went over by a couple characters
Jama.chpl:202 - thought I cleared out the commented-out Java references. I'll double check; if a couple jump out feel free to drop a line number or two so I can track them. :+1: 

I mostly skipped the stylistic suggestions for this PR. If ya'll definitely want those changes feel free to tell me if they are a hard-requirement for PR acceptance into this next release!

File: test/modules/standard/jama/TestMatrix.chpl (hunk 5 above)
Comment (lydia-duncan) [inline test/modules/standard/jama/TestMatrix.chpl]: You mean the Jama source?  I'd be fine with it directly under modules/standard.  The test code is probably best placed in a new folder test/modules/standard/LinearAlgebraJama

**Comment by lydia-duncan** [comment]
Overall, I like the comment block change.  I think I'd still remove the public domain copyright wording at the bottom of it (while keeping the authors intact) - since you reference the original port was public domain, I think that's sufficient.

I don't view the long lines as a showstopper since they aren't egregious, and I don't view the stylistic changes as a showstopper either.

Once the test code is moved, I'll make sure it runs just fine and then this ought to be good to go.  Thanks!

**Comment by ct-clmsn** [comment]
@lydia-duncan should be set!

**Comment by lydia-duncan** [comment]
Couple of test errors:

1) Since the name of the standard module file doesn't match the name of the top level module within it, the test file is having trouble finding the module you want.  To fix this, I'd change line 128's `module Jama {` to `module LinearAlgebraJama {` and the use in the test code to match this name.

2) When I do that, I get an error from the module itself:
$CHPL_HOME/modules/standard/LinearAlgebraJama.chpl:296: In function 'solve':
$CHPL_HOME/modules/standard/LinearAlgebraJama.chpl:308: error: 'k' undeclared (first use this function)

There's probably other errors behind this, so once those two are taken care of, would you mind running start_test over the test code?  Thanks

**Comment by ct-clmsn** [comment]
will do thanks!

**Comment by ct-clmsn** [comment]
The range cleanup worked well - I updated the test program's *.good file - all is looks good at the moment. Will work the documentation piece.

**Comment by lydia-duncan** [comment]
Passes for me as well, thanks!

**Comment by lydia-duncan** [comment]
I'm still seeing some Java for loops commented out (195, 199, 203, 258, 260, 262, 271, 273, 378, 391, 483, there's more but I think you can find the rest).  Since feature freeze is today, I'll not gate the commit on it, but those should get cleaned up with the documentation changes.

**Pull request merged**